File: src/pocketcloud/network/packet/CloudPacket.php
<?php

namespace pocketcloud\network\packet;

use pocketcloud\network\client\ServerClient;
use pocketcloud\network\packet\utils\PacketData;

abstract class CloudPacket  {

    private bool $encoded = false;

    public function encode(PacketData $packetData): void {
        if (!$this->encoded) {
            $this->encoded = true;
            $packetData->write((new \ReflectionClass($this))->getShortName());
            $this->encodePayload($packetData);
        } else throw new \RuntimeException("Packet: " . (new \ReflectionClass($this))->getShortName() . " is already encoded");
    }

    public function decode(PacketData $packetData): void {
        $packetData->readString();
        $this->decodePayload($packetData);
    }

    public function encodePayload(PacketData $packetData): void {}

    public function decodePayload(PacketData $packetData): void {}

    abstract public function handle(ServerClient $client);

    public function isEncoded(): bool {
        return $this->encoded;
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/CloudPacket.java
package de.pocketcloud.cloud.network.packet;

import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.packet.util.PacketData;
import lombok.Getter;
import lombok.Setter;

public abstract class CloudPacket implements Packet {
    
    private boolean encoded = false;
    private Long sentTimestamp = null;
    @Setter
    @Getter
    private int size = 0;
    
    @Override
    public void encode(PacketData packetData) {
        if (encoded) throw new RuntimeException("Packet " + getName() + " has already been encoded");
        encoded = true;
        packetData.write(getName()).write(sentTimestamp = System.currentTimeMillis());
        encodePayload(packetData);
    }
    
    @Override
    public void decode(PacketData packetData) {
        String packetName = packetData.readString();
        assert packetName != null;
        if (!packetName.equals(getName())) throw new RuntimeException("Packet name does not equal the actual class name? What have you done?");
        sentTimestamp = packetData.readLong();
        if (sentTimestamp == null) throw new RuntimeException("Packet data does not contain the actual sent timestamp? What have you done?");
        decodePayload(packetData);
    }

    public void sendPacket(ServerClient client) {
        if (!(this instanceof ClientboundPacket p)) return;
        client.sendPacket(p);
    }
    
    @Override
    public abstract void handle();

    @Override
    public final String getName() {
        return getClass().getSimpleName();
    }
    
    @Override
    public final boolean isEncoded() {
        return encoded;
    }
    
    @Override
    public final Long getSentTimestamp() {
        return sentTimestamp;
    }
}